Source/WebCore/dom/Document.cpp

```diff
@@ -10633,17 +10669,27 @@ DOMTimerHoldingTank& Document::domTimerHoldingTank()
 
 bool Document::isRunningUserScripts() const
 {
-    auto& top = topDocument();
-    return this == &top ? m_isRunningUserScripts : top.isRunningUserScripts();
+    if (isTopDocument())
+        return m_isRunningUserScripts;
+
+    if (RefPtr mainFrameDocument = protectedMainFrameDocument())
+        return mainFrameDocument->isRunningUserScripts();
+
+    LOG_ERROR_ONCE("Unable to properly calculate Document::isRunningUserScripts() without access to the main frame document (site isolation)");
+    return false;
 }
 
 void Document::setAsRunningUserScripts()
 {
-    Ref top = topDocument();
-    if (this == top.ptr())
+    if (isTopDocument()) {
         m_isRunningUserScripts = true;
+        return;
+    }
+
+    if (RefPtr mainFrameDocument = protectedMainFrameDocument())
+        mainFrameDocument->setAsRunningUserScripts();
     else
-        top->setAsRunningUserScripts();
+        LOG_ERROR_ONCE("Unable to properly set Document::setAsRunningUserScripts() without access to the main frame document (site isolation)");
```

Comment:
Do any of these print when running SiteIsolation API tests?

Reply:
They all do (but, just once per process)

The last time I caused issues with the tests with logging, it was unbounded logging - This was an attempt to resolve that?